java-checks/src/main/java/org/sonar/java/checks/RightCurlyBraceStartLineCheck.java

```diff
@@ -41,43 +61,153 @@
 @ActivatedByDefault
 @SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.READABILITY)
 @SqaleConstantRemediation("1min")
-public class RightCurlyBraceStartLineCheck extends SquidCheck<LexerlessGrammar> implements JavaCheck {
+public class RightCurlyBraceStartLineCheck extends SubscriptionBaseVisitor {
 
   @Override
-  public void init() {
-    subscribeTo(JavaPunctuator.RWING);
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.BLOCK);
   }
 
   @Override
-  public void visitNode(AstNode node) {
-    if (!isExcluded(node) && hasSomeCodeBefore(node)) {
-      getContext().createLineViolation(this, "Move this closing curly brace to the next line.", node);
+  public void visitNode(Tree tree) {
+    BlockTree blockTree = (BlockTree) tree;
+    List<StatementTree> body = blockTree.body();
+    SyntaxToken closeBraceToken = blockTree.closeBraceToken();
+    if (!body.isEmpty()) {
+      StatementTree lastStatement = body.get(body.size() - 1);
+      SyntaxToken lastToken = getLastToken(lastStatement);
+      if (lastToken.line() == closeBraceToken.line()) {
+        insertIssue(closeBraceToken);
+      }
+    } else if (closeBraceToken.line() == blockTree.openBraceToken().line()) {
+      insertIssue(closeBraceToken);
     }
   }
 
-  private static boolean hasSomeCodeBefore(AstNode node) {
-    Token previousToken = getPreviousToken(node);
-    return previousToken != null && previousToken.getLine() == node.getTokenLine();
+  private void insertIssue(SyntaxToken token) {
+    addIssue(token, "Move this closing curly brace to the next line.");
   }
 
-  private static boolean isExcluded(AstNode node) {
-    return node.getParent().is(Kind.NEW_ARRAY);
+  private SyntaxToken getLastToken(StatementTree lastStatement) {
+    StatementLastTokenFinder visitor = new StatementLastTokenFinder();
+    lastStatement.accept(visitor);
+    return visitor.lastToken;
   }
 
-  private static Token getPreviousToken(AstNode node) {
-    AstNode result = node.getPreviousAstNode();
+  private class StatementLastTokenFinder extends BaseTreeVisitor {
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Should org.sonar.java.checks.RightCurlyBraceStartLineCheck$StatementLastTokenFinder be a _static_ inner class?[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ASIC_INNER_SHOULD_BE_STATIC)